gossip/src/push_active_set.rs

```diff
@@ -0,0 +1,316 @@
+use {
+    crate::weighted_shuffle::WeightedShuffle,
+    indexmap::IndexMap,
+    rand::Rng,
+    solana_bloom::bloom::{AtomicBloom, Bloom},
+    solana_sdk::{native_token::LAMPORTS_PER_SOL, pubkey::Pubkey},
+    std::collections::HashMap,
+};
+
+// Equivalent to: get_stake_bucket(Some(&u64::MAX)) + 1.
+// Verified in test_get_stake_bucket.
+const NUM_PUSH_ACTIVE_SET_ENTRIES: usize = 36;
+
+pub(crate) struct PushActiveSet([PushActiveSetEntry; NUM_PUSH_ACTIVE_SET_ENTRIES]);
+
+// Keys are gossip nodes to push messages to.
+// Values are which origins the node has pruned.
+#[derive(Default)]
+struct PushActiveSetEntry(IndexMap</*node:*/ Pubkey, /*origins:*/ AtomicBloom<Pubkey>>);
+
+impl PushActiveSet {
+    #[cfg(debug_assertions)]
+    const MIN_NUM_BLOOM_ITEMS: usize = 512;
+    #[cfg(not(debug_assertions))]
+    const MIN_NUM_BLOOM_ITEMS: usize = crate::cluster_info::CRDS_UNIQUE_PUBKEY_CAPACITY;
+
+    pub(crate) fn get_nodes<'a>(
+        &'a self,
+        pubkey: &Pubkey,    // This node.
+        origin: &'a Pubkey, // CRDS value owner.
+        // If true forces gossip push even if the node has pruned the origin.
+        should_force_push: impl FnMut(&Pubkey) -> bool + 'a,
+        stakes: &HashMap<Pubkey, u64>,
+    ) -> impl Iterator<Item = &Pubkey> + 'a {
+        let stake = stakes.get(pubkey).min(stakes.get(origin));
```

Comment:
what's the reasoning for the min here instead of just using the `pubkey`'s stake? 

Is the goal to deprioritize messages from lowly staked `origin`'s as well?

Reply:
On the receiving end of gossip push messages:
https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L192
we keep enough incoming links to hit a `prune_stake_threshold` but that threshold is a function of the origin's stake:
https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L117-L120
So if the origin is unstaked or has low stake, we don't really care much about the stake of the incoming link.

So here on the push side, if the value is from an untaked or low staked `origin` then, instead of sampling nodes by stake, we may utilize the wider cluster to propagate it more efficiently.

> Is the goal to deprioritize messages from lowly staked origin's as well?

Not to deprioritize.
But instead of tunneling them through the high staked nodes (which would unnecessarily overload and create congestion at high staked nodes), use the wider cluster to propagate those.
For these values, the receiving node does not care if the incoming link has high stake or not, so we may sample the cluster more uniformly.

Reply:
> so the idea is to use nodes in the same bucket to propagate those messages from low staked origin
> but the messages will probably propagate through other lowly staked nodes in the same bucket?

Effectively
* For low staked origins we sample gossip nodes to push to uniformly, regardless of my node's stake.
* Similarly, if my node has low stake, then we sample gossip nodes to push to almost uniformly regardless of the origin's stake.
* But if my node has high stake and I receive a crds value from a high staked origin, then I would tend to push that value to another high staked node by giving those nodes higher sampling weight.

In other words, the only time the stakes matter is when both my node and the crds value origin have high stake. Otherwise gossip nodes to push to are sampled almost uniformly.

Again this is trying to mirror the similar logic on the receiving end:
https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L192
https://github.com/solana-labs/solana/blob/v1.14.11/gossip/src/crds_gossip_push.rs#L117-L120
where if my node has high stake, for crds values from a high staked origin, I would have a high `prune_stake_threshold` and therefore would prefer to receive these values from a high staked node.